astrophot/image/wcs.py

```diff
@@ -0,0 +1,759 @@
+from abc import ABC
+import torch
+import numpy as np
+
+from .. import AP_config
+from ..utils.conversions.units import deg_to_arcsec
+
+__all__ = ("WPCS", "PPCS", "WCS")
+
+deg_to_rad = np.pi / 180
+rad_to_deg = 180 / np.pi
+rad_to_arcsec = rad_to_deg * 3600
+arcsec_to_rad = deg_to_rad / 3600
+
+class WPCS:
+    """World to Plane Coordinate System in AstroPhot.
+
+    AstroPhot performs it's operations on a tangent plane to the
+    celestial sphere, this class handles projections between the sphere and the
+    tangent plane. It holds variables for the reference (RA,DEC) where
+    the tangent plane contacts the sphere, and the type of projection
+    being performed. Note that (RA,DEC) coordinates should always be
+    in degrees while the tangent plane is in arcsecs.
+
+    Attributes:
+      reference_radec: The reference (RA,DEC) coordinates in degrees where the tangent plane contacts the sphere.
+      reference_planexy: The reference tangent plane coordinates in arcsec where the tangent plane contacts the sphere.
+      projection: The projection system used to convert from (RA,DEC) onto the tangent plane. Should be one of: gnomonic (default), orthographic, steriographic
+
+    """
+
+    # Softening length used for numerical stability and/or integration stability to avoid discontinuities (near R=0)
+    softening = 1e-3
+    
+    default_reference_radec = (0,0)
+    default_reference_planexy = (0,0)
+    default_projection = "gnomonic"
+
+    def __init__(self, **kwargs):
+        self.projection = kwargs.get("projection", self.default_projection)
+        self.reference_radec = kwargs.get("reference_radec", self.default_reference_radec)
+        self.reference_planexy = kwargs.get("reference_planexy", self.default_reference_planexy)
+        
+        
+    def world_to_plane(self, world_RA, world_DEC=None):
+        """Take a coordinate on the world coordiante system, also called the
+        celesial sphere, (RA, DEC in degrees) and transform it to the
+        cooresponding tangent plane coordinate
+        (arcsec). Transformation is done based on the chosen
+        projection (default gnomonic) and reference positions. See the
+        :doc:`coordinates` documentation for more details on how the
+        transformation is performed.
+
+        """
+        
+        if world_DEC is None:
```

Comment:
Oh, this was a personal choice just to make the code easier to interface with. Sometimes you have RA and DEC as separate tensors and sometimes they are held together in a single tensor. I just added this so that both are acceptable ways to interface with the code and there is no ambiguity. If you pass both then the two tensors interpretation is obvious. If you just pass a single tensor then the first dimension is expanded into RA and DEC. So no matter what the shape of the tensors the behaviour of the function is predictable. I should mention it in the docstring though.